test/Test/Component/ForkTest.purs

```diff
@@ -0,0 +1,90 @@
+module Test.Component.ForkTest where
+
+import Prelude
+
+import Control.Coroutine as CR
+import Data.Foldable (traverse_)
+import Data.List ((:))
+import Data.List as L
+import Data.Maybe (Maybe(..))
+import Effect.Aff (Aff)
+import Effect.Aff as Aff
+import Effect.Class.Console (log)
+import Effect.Ref as Ref
+import Halogen as H
+import Test.Assert (assertEqual)
+import Test.TestDriver as TD
+
+type State = Maybe H.ForkId
+
+data Query a
+  = StartFork a
+  | KillFork a
+
+newtype Message = Message String
+
+derive newtype instance eqMessage :: Eq Message
+derive newtype instance showMessage :: Show Message
+
+component :: H.Component TD.TestRenderProduct Query Unit Message Aff
+component =
+  H.component
+    { initialState: const Nothing
+    , render: TD.render
+    , eval
+    , receiver: const Nothing
+    , initializer: Nothing
+    , finalizer: Nothing
+    }
+  where
+  eval :: Query ~> H.HalogenM State Query () Message Aff
+  eval = case _ of
+    StartFork a -> do
+      let
+        loop = do
+          H.liftAff $ Aff.delay (Aff.Milliseconds 100.0)
+          H.raise (Message "Progress")
+          loop
+      H.raise (Message "Starting")
+      fid <- H.fork loop
+      H.put (Just fid)
+      pure a
+    KillFork a -> do
+      H.get >>= traverse_ \fid -> do
+        H.put Nothing
+        H.kill fid
+      H.raise (Message "Killed")
+      pure a
+
+test :: Aff Unit
+test = do
+  io <- TD.runUI component unit
+
+  logRef <- H.liftEffect $ Ref.new L.Nil
+
+  io.subscribe $ CR.consumer \msg -> do
+    log (show msg)
+    H.liftEffect $ Ref.modify_ (msg : _) logRef
+    pure Nothing
+
+  io.query (H.action StartFork)
+  Aff.delay (Aff.Milliseconds 350.0)
+  io.query (H.action KillFork)
+
+  -- TODO: revisit this: why do we need to wait to receive `raise`d messages
+  -- from the component, if the `raise` occurs after any bind?
+  Aff.delay (Aff.Milliseconds 15.0)
```

Comment:
The Aff coroutines are forked and run in parallel separate from the main thread, so it may be just how the Aff runtime ends up interleaving things. If you want things to coordinate properly you should have a separate AVar that you take from when it gets the killed message.

Reply:
I did consider that. Tried to :duck: it with Christoph, there is definitely something a bit weird happening though - if the `raise` is the first thing in the `Kill` eval it works without this delay, if you even put `pure unit` in front of it then the delay is required. Basically it's the difference between the `raise` happening before the `io.query` completes, or after.

There's a chance it's specific to the driver IO behaviour too, since that has to jump through some hoops to produce the `subscribe` function for `HalogenIO`. I'm not too worried about it right now though, really I just wanted to check that forking and killing was working properly here.